### ISSUE TITLE
LFN support on Linux

### ISSUE THREAD

**Issue by Xeeynamo** [dosemu2/dosemu2#1817]
I am using a [toolchain in MS-DOS](https://github.com/Xeeynamo/wine-psyq/blob/master/Dockerfile.ubuntu-msdos) to build an [old videogame](https://github.com/Xeeynamo/sotn-decomp) that uses long file paths and file names that do not respect the MS-DOS' 8.3 file name syntax. The source codes are found on a Linux file system (EXT4) and they are mounted in `G:` by using the dosemu's flag `-K `. When passing a long file path to dosemu (eg. `TYPE asm/st/dre/data/0.rodata.s`) that is not recognised as an existing file.

I can see some references and previously created issues regarding LFN, but it does not work as expected on my side.

**Comment by stsp** [comment]
Try freecom instead of comcom32.
Symlink it as drive_c/command.com
and the problem should go away.

**Comment by Xeeynamo** [comment]
I got freecom from [FreeDOS' command.zip](https://github.com/FDOS/freecom/releases/tag/com085a) and installed it, but I still get the same problem:

```
C:\>type verylongfilename
type verylongfilename
File not found. - 'verylongfilename'
C:\>ver
ver

FreeCom version 0.85a - WATCOMC - XMS_Swap [Jul 10 2021 19:28:06]
```

I also followed the wiki to install [FreeDOS with `insfdusr`](https://github.com/dosemu2/dosemu2/wiki/Installing-various-DOSes#install-freedos-package) but it did not change the outcome either. Am I missing any particular step?

**Comment by andrewbird** [comment]
What does `dir /lfn` show you?

**Comment by Xeeynamo** [comment]
`dir /lfn` effectively shows to me the entire path. It did not work as dosemu is not currently booting straight into `C:\BIN\COMMAND.COM` where FreeCom is located as I have to do it manually. I followed your suggestion to create a symlink `ln /home/psyq/.dosemu/drive_c/BIN/COMMAND.COM /home/psyq/.dosemu/drive_c/COMMAND.COM` but running dosemu with `runuser -u psyq -- dosemu -quiet -dumb -K . -T` gives me no output as something is stuck. CTRL+C just kills dosemu as expected.

**Comment by stsp** [comment]
Its not stuck, its quiet.
Just what you asked for in a command line.
I checked `type verylongfilename` and
it works for me with freecom.

**Comment by stsp** [comment]
Please re-open when you have
more info. Since you never showed
the output of `dir /lfn` I suspect you
know that `verylongfilename` wasn't
there from the beginning.

**Comment by Xeeynamo** [comment]
> Its not stuck, its quiet. Just what you asked for in a command line. I checked `type verylongfilename` and it works for me with freecom.

Hi, I was building a system to replicate the issue. I confirm I still have the same problem. No LFS support with comcom32 and having freecom's `COMMAND.COM` into `~/.dosemu/drive_c` gets the command line stuck. If I use the `-T` flag to shell into dosemu and run `COMMAND.COM /P` then I have LFS support but it's a manual step it is strictly necessary to avoid in order to have a build system working. Down below you will find a Dockerfile that replicates the exact same environment I created. Notice how the very last command will get stuck.

Also I do not have permissions to re-open the current issue.

```
FROM ubuntu:latest AS freedos
WORKDIR /work
ENV DEBIAN_FRONTEND=noninteractive
ENV LANG=en_US

RUN apt-get update
RUN apt-get install -y curl unzip
RUN curl -L -o command.zip https://github.com/FDOS/freecom/releases/download/com085a/command.zip
RUN unzip command.zip

FROM ubuntu:latest
ENV DEBIAN_FRONTEND=noninteractive
ENV LANG=en_US

RUN apt-get update
RUN apt-get install -y software-properties-common
RUN add-apt-repository -y ppa:dosemu2/ppa
RUN apt-get update
RUN apt-get install -y dosemu2

RUN useradd psyq
RUN mkdir -p /home/psyq
RUN chown psyq /home/psyq/

# Install freecom
COPY --chown=psyq:psyq --from=freedos /work/BIN/COMMAND.COM /home/psyq/.dosemu/drive_c/

RUN printf '#!/bin/sh\nrunuser -u psyq -- dosemu -quiet -dumb -K . -E \"$@\"' >> /usr/bin/DOSEMU
RUN chmod +x /usr/bin/DOSEMU

# Testing
RUN touch verylongfilename
RUN DOSEMU type verylongfilename
```

**Comment by stsp** [comment]
Try this:
`dosemu -dumb -quiet -E "type verylongfilename" -I 'keystroke "\p10;\r\n\p10;\r\n"'`

**Comment by andrewbird** [comment]
Mmm, just thinking about this. As comcom32 doesn't support LFN at the moment you switched to freecom, but did you change the startup files to suit it? I'm wondering if it's waiting for you to set time and date etc.

**Comment by stsp** [comment]
This is exactly what the above command solves.

**Comment by andrewbird** [comment]
> This is exactly what the above command solves.

Was far too cryptic an incantation for me to realise that.

**Comment by stsp** [comment]
It just presses 2 enters with delays.
But I am going to remove /p for -E
automatically.

**Comment by Xeeynamo** [comment]
Interesting approach at sending the keystrokes. I will try it and post an update as soon as I am home.

**Comment by stsp** [comment]
I added automatic /p avoidance.

**Comment by Xeeynamo** [comment]
Hey, I noticed your recent commit and understood from there that `/p` was passed by default and it was causing the problem. So for now I resolved that part with `sed -i "s/shellhigh=.*/shellhigh\=command.com \/e:384 \/k %FDPP_AUTOEXEC%/g" /usr/share/dosemu/dosemu2-cmds-0.3/fdppconf.sys`, which essentially patches `fdppconf.sys` to remove that flag. Good news, if I do `TYPE verylongfilename` it now works. Bad news is that it is working only for the built-in commands of freecom. Any other MS-DOS executable will not be able to access any file that uses LFN. In short, I am back to square one with the issue. Am I missing anything?

**Comment by andrewbird** [comment]
> Any other MS-DOS executable will not be able to access any file that uses LFN

LFN support has to be compiled into each executable to be available. Some executables (DJGPP compiled) look at the environment to determine if to recognise LFNs. Try doing `set LFN=Y` before using the executable that gives you trouble. I note that you use the term `MS-DOS executable`, if you mean this literally then it's very unlikely that they have support for LFNs, as that was only provided by Win95 and beyond. On the other hand many FreeDOS executables have LFN support, most that do are compiled by DJGPP and so watch the `LFN` environment variable.

**Comment by Xeeynamo** [comment]
Thank you for your reply @andrewbird . I tried by setting the environment variable `LFN` to `Y` but the executable I am referring to still does not recognise the file(s) I try to pass into it. I was not aware that the LFN support needed to be built-in for each app. That pretty much solves the case as "it is not possible". I will try other options (eg. decompiling the app or statically recompile it). Thanks!

**Comment by andrewbird** [comment]
Yes there are different 'system calls' used for LFN. Bear in mind that the mapping from LFN to SFN is repeatable. So if your file is named 'verylongfilename.txt' its short name can be used by apps that can't access LFNs as 'VERYL~3G.TXT', use `DIR /LFN` to determine the mapping.

**Comment by stsp** [comment]
Repeatable in what sense?
I think ours is incompatible with windows...
So I suppose the one should use some
script to convert lfn to sfn, probably with
dir /lfn output? And then feed sfn to an app.
So this is far from being impossible, but
requires some work.

**Comment by andrewbird** [comment]
> Repeatable in what sense?

As in if I run Dosemu three times, providing that the external filesystem hasn't been modified in between times with clashing filenames, the mapping remains the same.

**Comment by stsp** [comment]
While definitely possible, I wouldnt
suggest to hard-code dosemu-specific
assumptions into someone's project.
You can map to sfn at run time.
And more so, I dont believe its actually
as stable as you suggest. If someone
creates more lfn files between dosemu
runs, I believe this can change even the
pre-existing sfns in case of clashes.

**Comment by stsp** [comment]
Maybe if we store sfn in xattrs then its
stable. But we do not. And I dont know
if we should.

**Comment by andrewbird** [comment]
> And more so, I dont believe its actually
as stable as you suggest. If someone
creates more lfn files between dosemu
runs, I believe this can change even the
pre-existing sfns in case of clashes.

I think that's what I said.

**Comment by stsp** [comment]
Indeed, sorry. :)
So its not really stable at the end.

**Comment by Xeeynamo** [comment]
> there are different 'system calls' used for LFN

By _system call_ you mean an interrupt? 'cause if it is easily patchable (eg. changing the interrupt address, which will involve a simple hex edit) then I will pursue that approach. I have very little knowledge on how the real-mode / DOS-era software is written, so some of my questions might sound commonplace.

**Comment by andrewbird** [comment]
> By system call you mean an interrupt?

Yes.

> if it is easily patchable

I wouldn't call it easy. Different interrupt services usually require different inputs in registers and may return different things. For instance probably the calls to open file will need to be changed. Looking at the RBIL for those calls will tell you what you need to do:
##### SFN open calls (file handle)
https://fd.lod.bz/rbil/interrup/dos_kernel/213d.html
https://fd.lod.bz/rbil/interrup/dos_kernel/216c00.html
##### LFN open call (file handle)
https://fd.lod.bz/rbil/interrup/dos_kernel/21716c.html

What existing calls the program makes will pretty much determine the likeliness of success, for instance if your program is really old it may use FCB file operations for which there is no direct LFN replacement.
##### SFN open call (FCB)
https://fd.lod.bz/rbil/interrup/dos_kernel/210f.html

And finally, you have to do all these changes whilst not altering the offsets of code and data within the executable. This usually means you'll have to start stealing space from somewhere (usually strings) to store new code.

So in summary, I think it's too hard, but I may not have the motivation that you have.

Is the program with the problem a common tool, or something niche? Perhaps there's an alternative available.

Let us know how you get on?
